src/libtest/lib.rs

```diff
@@ -1084,12 +1084,16 @@ impl MetricMap {
 /// elimination.
 ///
 /// This function is a no-op, and does not even read from `dummy`.
+#[cfg(not(all(target_os = "nacl", target_arch = "le32")))]
 pub fn black_box<T>(dummy: T) -> T {
     // we need to "use" the argument in some way LLVM can't
     // introspect.
     unsafe {asm!("" : : "r"(&dummy))}
     dummy
 }
+#[cfg(all(target_os = "nacl", target_arch = "le32"))]
+#[inline(never)]
+pub fn black_box<T>(dummy: T) -> T { dummy }
```

Comment:
This isn't a problem for NaCl which is native anyway, and thus supports asm!(); I think you're referring to PNaCl/JS.

I think you've made a fine point w.r.t. putting this in `rustc` so that it's impl can be based off of allow_asm, I just wonder why `#[inline(never)]` isn't enough on it's own (and thus wouldn't need to use asm in the first place). For example:
```rust
pub fn main() {
    #[inline(never)]
    fn black_box<T>(dummy: T) -> T {
        dummy
    }
    let two = black_box(2);
    let five = black_box(5);

    let ten = black_box(10);

    assert_eq!(two * five, ten);
}
```

This compiles to (optimized):
```llvm
...
; Function Attrs: uwtable
define internal void @_ZN4main20ha51c51acd3994054eaaE() unnamed_addr #0 {
entry-block:
  %ten = alloca i32, align 4
  %addr_of = alloca i32, align 4
  %0 = alloca %"2.core::fmt::Arguments", align 8
  %1 = alloca [2 x %"2.core::fmt::ArgumentV1"], align 8
  %2 = tail call fastcc i32 @_ZN4main9black_box21h10503287336754422282E(i32 2)
  %3 = tail call fastcc i32 @_ZN4main9black_box21h10503287336754422282E(i32 5)
  %4 = bitcast i32* %ten to i8*
  call void @llvm.lifetime.start(i64 4, i8* %4)
  %5 = tail call fastcc i32 @_ZN4main9black_box21h10503287336754422282E(i32 10)
  store i32 %5, i32* %ten, align 4
  %6 = mul i32 %3, %2
  %7 = bitcast i32* %addr_of to i8*
  call void @llvm.lifetime.start(i64 4, i8* %7)
  store i32 %6, i32* %addr_of, align 4
  %8 = icmp eq i32 %6, %5
  br i1 %8, label %next-block, label %then-block-61-
...
```

Is there some corner case I'm not aware of?